src/udev/udevadm-control.c

```diff
@@ -86,6 +86,19 @@ int control_main(int argc, char *argv[], void *userdata) {
         if (r < 0)
                 return log_error_errno(r, "Failed to initialize udev control: %m");
 
+        while ((c = getopt_long(argc, argv, "el:sSRp:m:t:Vh", options, NULL)) >= 0)
+                if (c == 't') {
+                        r = parse_sec(optarg, &timeout);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to parse timeout value '%s': %m", optarg);
+                        break;
+                }
+        optind = 0;
```

Comment:
hmm, wha's this? why move this early?

Reply:
Varlink calls are done from the arg. processing loop, hence the varlink timeout has to be set in advance. I don't like this much either, but the alternative is to postpone the actual work to after the arg. processing. That would be a more significant change which I'd rather keep out of this PR.

Reply:
why not just drop the timeout arg from udev_connection_init() and allow it to be set later?

then you don't have to do this ugly ahead-of time parsing

Reply:
or set some default initially, and update it if `t` is actually used?

Reply:
> why not just drop the timeout arg from udev_connection_init() and allow it to be set later?

That doesn't help. The issue this split is working around is that the timeout must be applied for all varlink calls, but that's not possible as long as the varlink calls are issued directly from the arg. processing loop. E.g., consider this command: `udevadm control -R -t 30`. It would call `io.systemd.service.Reload()` and _after that_ it would set timeout to 30s (which won't be used for anything). Or even more complex case: `udevadm control -R -t 10 -s -t 20 -S`. Here, each of the 3 varlink calls would be done with a different timeout. The only way to fix is to separate argument processing and execution, but I'd prefer to do that in a separate PR, as this one is large enough already.

Reply:
The only way to fix is to separate argument processing and execution, but I'd prefer to do that in a separate PR, as this one is large enough already.

Actually, I think I'll do that now and then rebase this PR on top of it.

Reply:
https://github.com/systemd/systemd/pull/29847